### PR TITLE
[Python Client] Fix handle complex schema

### LINKED ISSUE #7785: Failure of the Avro Serialization for Complex Types in the Python Client
**Describe the bug**
The [doc example](http://pulsar.apache.org/docs/en/2.6.0/client-libraries-python/#complex-types) fails with the following error:

> ValueError: <__main__.TestComplexSchema.test_serialize_complex_avro.<locals>.MySubRecord object at 0x7fabd361b7f0> (type <class '__main__.TestComplexSchema.test_serialize_complex_avro.<locals>.MySubRecord'>) do not match ['null', {'type': 'record', 'name': 'MySubRecord', 'fields': [{'name': 'x', 'type': ['null', 'int']}, {'name': 'y', 'type': ['null', 'long']}, {'name': 'z', 'type': ['null', 'string']}]}]

**To Reproduce**
Run the attached test

**Expected behavior**
The complex object should be serialized and deserialized to an equivalent object.

**Additional context**
The offending code is in the `AvroSchema` class in pulsar/pulsar-client-cpp/python/pulsar/schema/schema.py:
```python
class AvroSchema(Schema):
    def __init__(self, record_cls):
        super(AvroSchema, self).__init__(record_cls, _pulsar.SchemaType.AVRO,
                                         record_cls.schema(), 'AVRO')
        self._schema = record_cls.schema()

    def _get_serialized_value(self, x):
        if isinstance(x, enum.Enum):
            return x.name
        else:
            return x

    def encode(self, obj):
        self._validate_object_type(obj)
        buffer = io.BytesIO()
        m = {k: self._get_serialized_value(v) for k, v in obj.__dict__.items()}
        fastavro.schemaless_writer(buffer, self._schema, m)
        return buffer.getvalue()

    def decode(self, data):
        buffer = io.BytesIO(data)
        d = fastavro.schemaless_reader(buffer, self._schema)
        return self._record_cls(**d)
```

Tested on 2.5.2. Code didn't change in 2.6.0.
[pulsar_test.txt](https://github.com/apache/pulsar/files/5047658/pulsar_test.txt)

### PR DESCRIPTION
**Pull request by gaoran10**
Fixes #7785, #11221

### Motivation

Currently, the Pulsar python client couldn't handle complex schema, users using complex schema will encounter errors.

### Modifications

Fix AvroSchema encodes and decodes complex schema.
Fix JsonSchema decodes complex schema.

### Verifying this change

This change added tests and can be verified as follows:

  - *Encode and decode complex schema data*
  - *Produce and consume complex schema data*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (yes)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### CODE REVIEW

**Comment by Anonymitaet** [comment]
Thanks for your contribution. For this PR, do we need to update docs?

(The [PR template contains info about doc](https://github.com/apache/pulsar/blob/master/.github/PULL_REQUEST_TEMPLATE.md#documentation), which helps others know more about the changes. Can you provide doc-related info in this and future PR descriptions? Thanks)

File: pulsar-client-cpp/python/pulsar/schema/schema_avro.py [hunk 1]
```diff
@@ -45,14 +46,23 @@ def _get_serialized_value(self, x):
...
-            m = {k: self._get_serialized_value(v) for k, v in obj.__dict__.items()}
+            m = self.encode_dict(obj.__dict__)
             fastavro.schemaless_writer(buffer, self._schema, m)
             return buffer.getvalue()
 
+        def encode_dict(self, d: dict):
+            obj = {}
+            for k, v in d.items():
+                if isinstance(v, Record):
+                    obj[k] = self.encode_dict(v.__dict__)
```
Comment (merlimat) [inline pulsar-client-cpp/python/pulsar/schema/schema_avro.py:57]: This check could also be don in the `_get_serialized_value()` method:

```python
def _get_serialized_value(self, x):
            if isinstance(x, enum.Enum):
                return x.name
            elif isinstance(x, Record):
                self.encode_dict(x.__dict__)
            else:
                return x
```

File: pulsar-client-cpp/python/pulsar/schema/definition.py [hunk 2]
```diff
@@ -56,15 +57,20 @@ def _get_fields(cls, dct):
 
 class Record(with_metaclass(RecordMeta, object)):
 
-    def __init__(self, default=None, required_default=False, required=False, *args, **kwargs):
+    def __init__(self, default=None, required_default=False, required=False, decode=False, *args, **kwargs):
```
Comment (merlimat) [inline pulsar-client-cpp/python/pulsar/schema/definition.py:60]: It's not easy to understand what is the meaning of `decode=False|True` here. Also these arguments are exposed to the users (eg: marking a field as "required"), do we need to expose this `decode` too?

File: pulsar-client-cpp/python/pulsar/schema/schema_avro.py (hunk 1 above)
Comment (gaoran10) [reply pulsar-client-cpp/python/pulsar/schema/schema_avro.py]: I fixed this. PTAL

File: pulsar-client-cpp/python/pulsar/schema/definition.py (hunk 2 above)
Comment (gaoran10) [reply pulsar-client-cpp/python/pulsar/schema/definition.py]: I think this is not necessary, I remove the param `decode`.

**Review by merlimat — APPROVED** [review]

**Review by sijie — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Comment by codelipenghui** [comment]
@gaoran10 Please rebase to the master branch to make the CI gets passed.

**Pull request merged**